File: jetbrains-core/tst/software/aws/toolkits/jetbrains/ui/wizard/SchemaCodeGenUtilsTest.kt
// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package software.aws.toolkits.jetbrains.ui.wizard

import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

// TODO: This is fragile. Very fragle. But it is necessary to get Schemas service launched, and we've evaluated all other tradeoffs
// This will be done on the server-side as soon as we can, but for now the client needs to do this
class SchemaCodeGenUtilsTest {

    @Test
    public fun awsEventPackageName() {
        assertThat(SchemaCodeGenUtils.buildSchemaPackageName("aws.ec2.EC2InstanceStateChangeNotificationEvent"))
            .isEqualTo("schema.aws.ec2.ec2instancestatechangenotificationevent")
        assertThat(SchemaCodeGenUtils.buildSchemaPackageName("aws.ec2@EC2InstanceStateChangeNotificationEvent"))
            .isEqualTo("schema.aws.ec2.ec2instancestatechangenotificationevent")
    }

    @Test
    public fun partnerEventPackageName() {
        assertThat(SchemaCodeGenUtils.buildSchemaPackageName("aws.partner-mongodb.com/1234567-tickets@TicketCreated"))
            .isEqualTo("schema.aws.partner.mongodb_com_1234567_tickets.ticketcreated")
        assertThat(SchemaCodeGenUtils.buildSchemaPackageName("aws.partner-zendesk.com/some#other#special#chars@MyEvent"))
            .isEqualTo("schema.aws.partner.zendesk_com_some_other_special_chars.myevent")
        assertThat(SchemaCodeGenUtils.buildSchemaPackageName("aws.partner-pagerduty.com@YouGotPaged"))
            .isEqualTo("schema.aws.partner.pagerduty_com.yougotpaged")
    }

    @Test
    public fun customerUploadedEventPackageName() {
        assertThat(SchemaCodeGenUtils.buildSchemaPackageName("MyEvent"))
            .isEqualTo("schema.myevent")
        assertThat(SchemaCodeGenUtils.buildSchemaPackageName("MyEvent.Special#Characters${'$'}etc"))
            .isEqualTo("schema.myevent_special_characters_etc")
        assertThat(SchemaCodeGenUtils.buildSchemaPackageName("MyEvent@Discriminator"))
            .isEqualTo("schema.myevent.discriminator")
        assertThat(SchemaCodeGenUtils.buildSchemaPackageName("MyEvent@Discriminator@Another"))
            .isEqualTo("schema.myevent.discriminator.another")
    }

    @Test
    public fun formatIdentifier() {
        assertThat(SchemaCodeGenUtils.IdentifierFormatter.toValidIdentifier("Review.created"))
            .isEqualTo("Review_created")
    }
}

File: src/test/eventSchemas/model/schemaCodeGenUtils.test.ts
/*!
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

import * as assert from 'assert'
import { IdentifierFormatter, SchemaCodeGenUtils } from '../../../eventSchemas/models/schemaCodeGenUtils'

describe('SchemaCodeGenUtils', async () => {
    const testScenarios = [
        {
            scenario: 'builds the "aws event" package name',
            input: 'aws.ec2.EC2InstanceStateChangeNotificationEvent',
            expectedResult: 'schema.aws.ec2.ec2instancestatechangenotificationevent'
        },
        {
            scenario: 'builds the "aws event" package name containing package seperator @',
            input: 'aws.ec2@EC2InstanceStateChangeNotificationEvent',
            expectedResult: 'schema.aws.ec2.ec2instancestatechangenotificationevent'
        },
        {
            scenario: 'builds the "partner event" package name containing slash,numbers,dash and package seperator @',
            input: 'aws.partner-mongodb.com/1234567-tickets@TicketCreated',
            expectedResult: 'schema.aws.partner.mongodb_com_1234567_tickets.ticketcreated'
        },
        {
            scenario: 'builds the "partner event" package name containing other special characters',
            input: 'aws.partner-zendesk.com/some#other?special;chars@MyEvent',
            expectedResult: 'schema.aws.partner.zendesk_com_some_other_special_chars.myevent'
        },
        {
            scenario: 'builds the "partner event" package name containing package seperator @ only',
            input: 'aws.partner-pagerduty.com@YouGotPaged',
            expectedResult: 'schema.aws.partner.pagerduty_com.yougotpaged'
        },
        {
            scenario: 'builds the "custom event" package name containing no special characters',
            input: 'MyEvent',
            expectedResult: 'schema.myevent'
        },
        {
            scenario: 'builds the "custom event" package name containing special characters',
            input: 'MyEvent.Special#Characters$etc',
            expectedResult: 'schema.myevent_special_characters_etc'
        },
        {
            scenario: 'builds the "custom event" package name containing package seperator',
            input: 'MyEvent@Discriminator',
            expectedResult: 'schema.myevent.discriminator'
        },
        {
            scenario: 'builds the "custom event" package name containing multiple package seperators',
            input: 'MyEvent@Discriminator@Another',
            expectedResult: 'schema.myevent.discriminator.another'
        }
    ]

    describe('buildSchemaPackageName', async () => {
        testScenarios.forEach(test => {
            it(test.scenario, async () => {
                const codeGen = new SchemaCodeGenUtils()
                const result = codeGen.buildSchemaPackageName(test.input)
                assert.strictEqual(result, test.expectedResult, 'Invalid package name returned')
            })
        })
    })
})

describe('IdentifierFormatter', async () => {
    describe('toValidIdentifier', async () => {
        it('replaces invalid identifier characters with underscore', async () => {
            assert.strictEqual(IdentifierFormatter.toValidIdentifier('Review.created?today'), 'Review_created_today')
        })

        it('replaces package seperator @ with dot', async () => {
            assert.strictEqual(IdentifierFormatter.toValidIdentifier('Review@created'), 'Review.created')
        })
    })
})
